File: Data Structures & Algorithms/invert-a-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        node = root
        que = deque([node])
        while que:
            node = que.popleft()
            if not node:
                continue
            node.left, node.right = node.right, node.left
            que.extend([node.left, node.right])
        return root

        

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-3.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {number[]} preorder
     * @param {number[]} inorder
     * @return {TreeNode}
     */
    buildTree(preorder, inorder) {
        if (preorder.length == 0 || inorder.length == 0) {
            return null;
        }

        const root = new TreeNode(preorder[0]);

        let j = 0;

        while (j < inorder.length) {
            if (inorder[j] == root.val) break
            j ++;
        }

        const leftChild = this.buildTree(
            preorder.slice(1, j + 1), 
            inorder.slice(0, j)
        );

        const rightChild = this.buildTree(
            preorder.slice(j + 1, preorder.length),
            inorder.slice(j + 1, inorder.length)
        );

        root.left = leftChild;
        root.right = rightChild;

        return root;
    }
}
